sip/sip-009-analysis-cost-assessment.md

```diff
@@ -0,0 +1,333 @@
+# SIP 009 Clarity Parsing and Analysis Cost Assessment
+
+## Preamble
+
+Title: Clarity Parsing and Analysis Cost Assessment
+
+Author: Aaron Blankstein <aaron@blockstack.com>
+
+Status: Draft
+
+Type: Standard
+
+Created: 03/05/2020
+
+License: BSD 2-Clause
+
+# Abstract
+
+This document describes the measured costs and asymptotic costs
+assessed for parsing Clarity code into an abstract syntax tree (AST)
+and the static analysis of that Clarity code (type-checking and
+read-only enforcement). This will not specify the _constants_
+associated with those asymptotic cost functions. Those constants will
+necessarily be measured via benchmark harnesses and regression
+analyses.
+
+# Measurements for Execution Cost
+
+The cost of analyzing Clarity code is measured using the same 5 categories
+described in SIP-006 for the measurement of execution costs:
+
+1. Runtime cost: captures the number of cycles that a single
+   processor would require to process the Clarity block. This is a
+   _unitless_ metric, so it will not correspond directly to cycles,
+   but rather is meant to provide a basis for comparison between
+   different Clarity code blocks.
+2. Data write count: captures the number of independent writes
+   performed on the underlying data store (see SIP-004).
+3. Data read count: captures the number of independent reads
+   performed on the underlying data store.
+4. Data write length: the number of bytes written to the underlying
+   data store.
+5. Data read length: the number of bytes read from the underlying
+   data store.
+
+Importantly, these costs are used to set a _block limit_ for each
+block.  When it comes to selecting transactions for inclusion in a
+block, miners are free to make their own choices based on transaction
+fees, however, blocks may not exceed the _block limit_. If they do so,
+the block is considered invalid by the network --- none of the block's
+transactions will be materialized and the leader forfeits all rewards
+from the block.
+
+Costs for static analysis are assessed during the _type check_ pass.
+The read-only and trait-checking passes perform work which is strictly
+less than the work performed during type checking, and therefore, the
+cost assessment can safely fold any costs that would be incurred during
+those passes into the type checking pass.
+
+# Common Analysis Metrics and Costs
+
+## AST Parsing
+
+The Clarity parser has a runtime that is linear with respect to the Clarity
+program length.
```

Comment:
This is true with the regex library we're using?

Reply:
It's true for the regex patterns we are using, but not general regexes.

Reply:
A related observation: While the tokenizer is linear, it searches across the regex patterns for the matching token, so **the lexical analysis runtime also depends on the type of tokens in the input.**

However, basic s-expressions can be tokenized with a single character look-ahead, and except for the tokens starting with a `'` the [Clarity lexical grammar](https://github.com/blockstack/blockstack-core/blob/7d162419f0bd77a4273a158a989669ee5f8fe9c0/src/vm/ast/parser/mod.rs#L76) is LL(2) allowing `0x` hexstrings and `u` for unsigned integers.

**This allows for a faster and more predictable tokenizer runtime** e.g by using an FSM with a O(1) lookup from character to token.

Reply:
Yep -- that's true @njordhov. For the purposes of the cost tracking, this doesn't really affect too much, because the number of regex patterns is fixed and relatively small. Using something like a parser generator to construct an FSM for lexing the grammar would speed things up, but I think we can leave that as a future project for now.